### ISSUE TITLE
[JENKINS-34210] Memory leak in Jenkins frontend when updating executors

### ISSUE THREAD

**Issue by jenkins-infra-bot** [timja-org/jenkins-gh-issues-poc-11-15#5687]
<p>In ./war/src/main/webapp/scripts/hudson-behavior.js there is a memory leak triggered by replacing the HTML for the executors panel with fresh content. On installations with a huge list of executors, Chrome can increase in memory usage with over 200MBs every 10minutes by just having the Jenkins dashboard open.</p>

<pre>
<span class="code-comment">// refresh a part of the HTML specified by the given ID,
</span><span class="code-comment">// by using the contents fetched from the given URL.
</span>function refreshPart(id,url) {
    <span class="code-keyword">var</span> f = function() {
<span class="code-keyword">if</span>(isPageVisible()) {
    <span class="code-keyword">new</span> Ajax.Request(url, {
onSuccess: function(rsp) {
    <span class="code-keyword">var</span> hist = $(id);
    <span class="code-keyword">if</span> (hist == <span class="code-keyword">null</span>) {
console.log(<span class="code-quote">"There's no element that has ID of "</span> + id);
<span class="code-keyword">return</span>;
    }
    <span class="code-keyword">var</span> p = hist.up();

    <span class="code-keyword">var</span> div = document.createElement(<span class="code-quote">'div'</span>);
    div.innerHTML = rsp.responseText;

    <span class="code-keyword">var</span> node = $(div).firstDescendant();
    p.replaceChild(node, hist);
 ########## hist cannot be garbage collected after the replaceChild!
 ########## If I comment out applySubtree(node) the memory leak disappears
 ########## So perhaps some event listeners stopping the GC from cleaning up?
    Behaviour.applySubtree(node);
    layoutUpdateCallback.call();

    <span class="code-keyword">if</span>(isRunAsTest) <span class="code-keyword">return</span>;
    refreshPart(id,url);
}
    });    
} <span class="code-keyword">else</span> {
    <span class="code-comment">// Reschedule
</span>    <span class="code-keyword">if</span>(isRunAsTest) <span class="code-keyword">return</span>;
    refreshPart(id,url);
}

    };
    <span class="code-comment">// <span class="code-keyword">if</span> run as test, just <span class="code-keyword">do</span> it once and <span class="code-keyword">do</span> it now to make sure it's working,
</span>    <span class="code-comment">// but don't repeat.
</span>    <span class="code-keyword">if</span>(isRunAsTest) f();
    <span class="code-keyword">else</span>    window.setTimeout(f, 5000);
}

</pre>

---
<details><summary><i>Originally reported by <img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/gustafl.png" title="gustafl's avatar" /> <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=gustafl">gustafl</a>, imported from: <a href="https://issues.jenkins.io/browse/JENKINS-34210" target="_blank">Memory leak in Jenkins frontend when updating executors</a></i></summary>
<i><ul>
<li><b>status</b>: Open
<li><b>priority</b>: Critical
<li><b>component(s)</b>: core
<li><b>label(s)</b>: core, jenkins, memory-leak, performance
<li><b>resolution</b>: Unresolved
<li><b>votes</b>: 8
<li><b>watchers</b>: 14
<li><b>imported</b>: 2025-11-16
</ul></i>
<details><summary>Raw content of original issue</summary>

<pre>
<p>In ./war/src/main/webapp/scripts/hudson-behavior.js there is a memory leak triggered by replacing the HTML for the executors panel with fresh content. On installations with a huge list of executors, Chrome can increase in memory usage with over 200MBs every 10minutes by just having the Jenkins dashboard open.</p>

<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java"><span class="code-comment">// refresh a part of the HTML specified by the given ID,
</span><span class="code-comment">// by using the contents fetched from the given URL.
</span>function refreshPart(id,url) {
    <span class="code-keyword">var</span> f = function() {
        <span class="code-keyword">if</span>(isPageVisible()) {
            <span class="code-keyword">new</span> Ajax.Request(url, {
                onSuccess: function(rsp) {
                    <span class="code-keyword">var</span> hist = $(id);
                    <span class="code-keyword">if</span> (hist == <span class="code-keyword">null</span>) {
                        console.log(<span class="code-quote">"There's no element that has ID of "</span> + id);
                        <span class="code-keyword">return</span>;
                    }
                    <span class="code-keyword">var</span> p = hist.up();

                    <span class="code-keyword">var</span> div = document.createElement(<span class="code-quote">'div'</span>);
                    div.innerHTML = rsp.responseText;

                    <span class="code-keyword">var</span> node = $(div).firstDescendant();
                    p.replaceChild(node, hist);
 ########## hist cannot be garbage collected after the replaceChild!
 ########## If I comment out applySubtree(node) the memory leak disappears
 ########## So perhaps some event listeners stopping the GC from cleaning up?
                    Behaviour.applySubtree(node);
                    layoutUpdateCallback.call();

                    <span class="code-keyword">if</span>(isRunAsTest) <span class="code-keyword">return</span>;
                    refreshPart(id,url);
                }
            });    
        } <span class="code-keyword">else</span> {
            <span class="code-comment">// Reschedule
</span>            <span class="code-keyword">if</span>(isRunAsTest) <span class="code-keyword">return</span>;
            refreshPart(id,url);
        }
        
    };
    <span class="code-comment">// <span class="code-keyword">if</span> run as test, just <span class="code-keyword">do</span> it once and <span class="code-keyword">do</span> it now to make sure it's working,
</span>    <span class="code-comment">// but don't repeat.
</span>    <span class="code-keyword">if</span>(isRunAsTest) f();
    <span class="code-keyword">else</span>    window.setTimeout(f, 5000);
}

</pre>
</div></div></pre>
</details>
</details>
<ul><li><i>environment</i>: <code>Chrome (Ubuntu, Debian), Firefox.</code></li></ul>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/danielbeck.png" title="danielbeck's avatar" /> danielbeck's <a href="https://issues.jenkins.io/browse/JENKINS-34210?focusedId=254854&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-254854">comment</a>:</i></summary>


<details><summary><sub><i>Raw content of original comment:</i></sub></summary>

<pre>
<p>Any idea whether this is a recent change?</p>
</pre>
</details>

</details>
<p>Any idea whether this is a recent change?</p>
